### ISSUE TITLE
support async await w/o including regenerator

### ISSUE THREAD

**Issue by tomwayson** [Esri/hub.js#250]
We want to be able to use `async`/`await` in this library. It's 2020 ffs.

The problem is that TS compiles our ESM build to [ES5](https://github.com/Esri/hub.js/blob/7131d91869a89064cd7c7eb73b148b8d5a6d8fe0/tsconfig.json#L4), which means [it would include a lot of generator code](https://mariusschulz.com/blog/compiling-async-await-to-es3-es5-in-typescript#compiling-async-await-to-es3-es5). As far as I understand this would be included redundantly in each package.

I like the idea of a TS equivalent of [fast-async](https://github.com/MatAtBread/fast-async) - a babel plugin that transforms the statements at build time to promises, but I'm not aware of a way to do that w/ TS.

Another alternative is that we generate a 2017 ESM build alongside the ES5 build and configure ember-auto-import in the Hub ember app to import that build. The ES5 code w/ the generator would still go into the UMD builds. As long as the Hub never consumes those, I don't care how much crap we stuff in them.

**Comment by tomwayson** [comment]
Well, it looks like we already have `__awaiter` and `__generator` in our code:

```
12 results - 6 files

packages/initiatives/dist/esm/activate.js:
   3: import { __assign, __awaiter, __generator } from "tslib";
  40:         return __generator(this, function (_a) {

packages/initiatives/dist/esm/remove.js:
   1: import { __assign, __awaiter, __generator } from "tslib";
  33:         return __generator(this, function (_a) {

packages/search/dist/esm/ago/compute-items-facets.js:
   1: import { __assign, __awaiter, __generator } from "tslib";
  30:         return __generator(this, function (_a) {

packages/search/dist/esm/ago/get-items.js:
   1: import { __assign, __awaiter, __generator } from "tslib";
  10:         return __generator(this, function (_a) {

packages/search/dist/esm/ago/search.js:
   3: import { __awaiter, __generator } from "tslib";
  18:         return __generator(this, function (_a) {

packages/sites/dist/esm/domains/_check-status-and-parse-json.js:
  1: import { __awaiter, __generator } from "tslib";
  9:         return __generator(this, function (_a) {
```

At first I thought it was from this recent change:

https://github.com/Esri/hub.js/blame/7131d91869a89064cd7c7eb73b148b8d5a6d8fe0/tsconfig.json#L8-L9

That ^^^ is worrisome for me on other levels. What does that do? Does that actually include polyfills for those, or does it just tell TS to chill out if the target is ES5 but somebody uses `[].includes()`?

Anyway, it looks like we've had `async`/`await` for a long time:

https://github.com/Esri/hub.js/blame/master/packages/initiatives/src/remove.ts#L42

One thing I was wrong about above is that b/c we use tslib, at least the implementations of the above code are **not** included in our ESM builds, and therefore [not included redundantly in each package](https://github.com/Microsoft/tslib#tslib).

Those functions along w/ all the other `tslib` fns used by each packge _are_ included in the UMD builds. Again, we don't care about that - as long as we don't bring the UMD builds into the Hub Ember app.

However, when I look into my local copy of vendor.js I see that `__generator` is declared at least 5 times. I'll have to investigate why. If it was only in there once, I guess I could live w/ that even though it's totally redundant to the babel regenerator code that's included in the ember build (and who knows how many times _that_ is redundantly included).

Ultimately I think our best hope is to output a non-ES5 build and try consuming that in the Ember app.

**Comment by tomwayson** [comment]
So I added these libs in a new ember app, and only got one copy of the `tslib` helpers. I suspect the issue has something to do w/ how these dependencies are used by multiple packages w/in the Hub monorepo.

When I opened this issue, I assumed that the npm ecosytem had moved beyond publishing ES5 in `modules` and I just missed the memo, but after reading https://jasonformat.com/enabling-modern-js-on-npm/ it sounds like ES5 is still the de facto standard. 😦 

In thinking through my above suggestion to also publish ES2017 and configure the Hub app to use that, I realized we'd not only have to configure ember-auto-import to use the ES2017 but _also_ configure webpack to transpile it, which would increase complexity and duration of builds. We might be able to get away w/ it since _most_ of our app is built outside of webpack, but it's risky.

After talking w/ @dbouwman and @drewctate I think the best course of action is to do a complete 180 from my original proposal and continue to prohibit the use of async/await in this library.
